File: sdk/python/tennetctl/iam.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ._transport import Transport


class _ReadResource:
    """Read-only resource wrapper for IAM list/get endpoints."""

    def __init__(self, transport: Transport, path: str) -> None:
        self._t = transport
        self._path = path  # e.g. "/v1/users"

    async def list(self, **filters: Any) -> list[dict]:
        params = {k: v for k, v in filters.items() if v is not None}
        data = await self._t.request("GET", self._path, params=params or None)
        return data if isinstance(data, list) else list(data or [])

    async def get(self, item_id: str) -> dict:
        return await self._t.request("GET", f"{self._path}/{item_id}")


class IAM:
    """Read-only helpers for IAM resources. Mutation goes through admin UI / direct endpoints."""

    def __init__(self, transport: Transport) -> None:
        self.users = _ReadResource(transport, "/v1/users")
        self.orgs = _ReadResource(transport, "/v1/orgs")
        self.workspaces = _ReadResource(transport, "/v1/workspaces")
        self.roles = _ReadResource(transport, "/v1/roles")
        self.groups = _ReadResource(transport, "/v1/groups")

File: sdk/typescript/src/iam.ts
import type { Transport } from "./transport.js";

class ReadResource {
  constructor(
    private readonly t: Transport,
    private readonly path: string,
  ) {}

  async list(filters: Record<string, string | number | boolean | undefined> = {}): Promise<
    Array<Record<string, unknown>>
  > {
    const params: Record<string, string | number | boolean | undefined> = {};
    for (const [k, v] of Object.entries(filters)) {
      if (v !== undefined) params[k] = v;
    }
    const data = await this.t.request<Array<Record<string, unknown>>>(
      "GET",
      this.path,
      { params: Object.keys(params).length ? params : undefined },
    );
    return Array.isArray(data) ? data : [];
  }

  async get(id: string): Promise<Record<string, unknown>> {
    return this.t.request("GET", `${this.path}/${id}`);
  }
}

export class IAM {
  readonly users: ReadResource;
  readonly orgs: ReadResource;
  readonly workspaces: ReadResource;
  readonly roles: ReadResource;
  readonly groups: ReadResource;

  constructor(t: Transport) {
    this.users = new ReadResource(t, "/v1/users");
    this.orgs = new ReadResource(t, "/v1/orgs");
    this.workspaces = new ReadResource(t, "/v1/workspaces");
    this.roles = new ReadResource(t, "/v1/roles");
    this.groups = new ReadResource(t, "/v1/groups");
  }
}
